# 01 Data Understanding

## Melbourne Metro Rail Patronage Analytics Project

This notebook explores the structure, quality, and consistency of the raw Melbourne metropolitan train station patronage datasets before data cleaning and consolidation.

## 1. Objective

The purpose of this notebook is to:

- Load and inspect the raw patronage datasets
- Understand available columns and data types
- Review row and column counts
- Check missing values and duplicates
- Compare file structures across years
- Identify early data quality issues before creating the master dataset

In [3]:
# Import Libraries
import pandas as pd
from pathlib import Path

In [4]:
# Set Project Paths
PROJECT_ROOT = Path("..")

RAW_DATA_DIR = PROJECT_ROOT / "data" / "raw"
PROCESSED_DATA_DIR = PROJECT_ROOT / "data" / "processed"

RAW_DATA_DIR

PosixPath('../data/raw')

In [5]:
# List Available Raw Files
raw_files = sorted(RAW_DATA_DIR.glob("*.csv"))

raw_files

[PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2018_2019.csv'),
 PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2019_2020.csv'),
 PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2020_2021.csv'),
 PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2021_2022.csv'),
 PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2022_2023.csv'),
 PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2023_2024.csv'),
 PosixPath('../data/raw/annual_metropolitan_train_station_entries_fy_2024_2025.csv')]

In [6]:
# Create File Inventory
file_inventory = []

for file in raw_files:
    df_temp = pd.read_csv(file)
    
    file_inventory.append({
        "file_name": file.name,
        "rows": df_temp.shape[0],
        "columns": df_temp.shape[1],
        "column_names": list(df_temp.columns)
    })

inventory_df = pd.DataFrame(file_inventory)

inventory_df[["file_name", "rows", "columns"]]

,file_name,rows,columns
0,annual_metropolitan_train_station_entries_fy_2...,222,16
1,annual_metropolitan_train_station_entries_fy_2...,222,16
2,annual_metropolitan_train_station_entries_fy_2...,222,16
3,annual_metropolitan_train_station_entries_fy_2...,222,16
4,annual_metropolitan_train_station_entries_fy_2...,223,16
5,annual_metropolitan_train_station_entries_fy_2...,222,16
6,annual_metropolitan_train_station_entries_fy_2...,222,16


In [7]:
# Load Latest Dataset
latest_file = RAW_DATA_DIR / "annual_metropolitan_train_station_entries_fy_2024_2025.csv"

df = pd.read_csv(latest_file)

df.head()

,Fin_year,Stop_ID,Stop_name,Stop_lat,Stop_long,Pax_annual,Pax_weekday,Pax_norm_weekday,Pax_sch_hol_weekday,Pax_Saturday,Pax_Sunday,Pax_pre_AM_peak,Pax_AM_peak,Pax_interpeak,Pax_PM_peak,Pax_PM_late
0,FY24-25,15351,Sunbury,-37.579091,144.727319,691050,2200,2250,1950,1450,950,450,850,450,400,100
1,FY24-25,15353,Diggers Rest,-37.627017,144.719922,165600,550,600,450,200,150,100,300,100,50,50
2,FY24-25,19827,Stony Point,-38.374235,145.221837,6950,50,50,50,50,50,50,50,50,50,50
3,FY24-25,19828,Crib Point,-38.366123,145.204043,7900,50,50,50,50,50,50,50,50,50,50
4,FY24-25,19829,Morradoo,-38.354033,145.189603,3400,50,50,50,50,50,50,50,50,50,50


In [8]:
# Check shape
df.shape

(222, 16)

The latest dataset contains 222 rows and 16 columns.

In [9]:
# Inspect Column Names
df.columns.tolist()

['Fin_year',
 'Stop_ID',
 'Stop_name',
 'Stop_lat',
 'Stop_long',
 'Pax_annual',
 'Pax_weekday',
 'Pax_norm_weekday',
 'Pax_sch_hol_weekday',
 'Pax_Saturday',
 'Pax_Sunday',
 'Pax_pre_AM_peak',
 'Pax_AM_peak',
 'Pax_interpeak',
 'Pax_PM_peak',
 'Pax_PM_late']

In [10]:
# Inspect Data Types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 222 entries, 0 to 221
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Fin_year             222 non-null    str    
 1   Stop_ID              222 non-null    int64  
 2   Stop_name            222 non-null    str    
 3   Stop_lat             222 non-null    float64
 4   Stop_long            222 non-null    float64
 5   Pax_annual           222 non-null    int64  
 6   Pax_weekday          222 non-null    int64  
 7   Pax_norm_weekday     222 non-null    int64  
 8   Pax_sch_hol_weekday  222 non-null    int64  
 9   Pax_Saturday         222 non-null    int64  
 10  Pax_Sunday           222 non-null    int64  
 11  Pax_pre_AM_peak      222 non-null    int64  
 12  Pax_AM_peak          222 non-null    int64  
 13  Pax_interpeak        222 non-null    int64  
 14  Pax_PM_peak          222 non-null    int64  
 15  Pax_PM_late          222 non-null    int64  
dtypes

In [11]:
# Summary Statistics
df.describe(include="all")

,Fin_year,Stop_ID,Stop_name,Stop_lat,Stop_long,Pax_annual,Pax_weekday,Pax_norm_weekday,Pax_sch_hol_weekday,Pax_Saturday,Pax_Sunday,Pax_pre_AM_peak,Pax_AM_peak,Pax_interpeak,Pax_PM_peak,Pax_PM_late
count,222,222.000000,222,222.000000,222.000000,2.220000e+02,222.000000,222.000000,222.000000,222.000000,222.000000,222.000000,222.000000,222.000000,222.000000,222.000000
unique,1,NaN,222,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,FY24-25,NaN,Sunbury,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,222,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,21510.909910,NaN,-37.852946,145.046270,8.039671e+05,2569.369369,2666.216216,2217.117117,1554.504505,1123.423423,154.279279,738.738739,575.900901,982.657658,238.738739
std,NaN,6568.383093,NaN,0.140689,0.143188,1.896382e+06,6033.303768,6200.006065,5423.601413,3808.578965,2819.725443,160.435418,613.007004,1292.536126,3579.232919,958.155862
min,NaN,15351.000000,NaN,-38.374235,144.661118,2.500000e+03,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000
25%,NaN,19880.250000,NaN,-37.900124,144.960580,2.520250e+05,800.000000,850.000000,650.000000,462.500000,350.000000,50.000000,350.000000,150.000000,150.000000,50.000000
50%,NaN,19937.500000,NaN,-37.826147,145.035954,4.179500e+05,1350.000000,1400.000000,1150.000000,825.000000,550.000000,100.000000,600.000000,300.000000,300.000000,50.000000
75%,NaN,19995.750000,NaN,-37.771140,145.121494,6.963375e+05,2137.500000,2250.000000,1800.000000,1337.500000,950.000000,200.000000,900.000000,487.500000,550.000000,100.000000


In [12]:
# Missing Value Check
missing_values = df.isna().sum().reset_index()
missing_values.columns = ["column", "missing_count"]

missing_values

,column,missing_count
0,Fin_year,0
1,Stop_ID,0
2,Stop_name,0
3,Stop_lat,0
4,Stop_long,0
5,Pax_annual,0
6,Pax_weekday,0
7,Pax_norm_weekday,0
8,Pax_sch_hol_weekday,0
9,Pax_Saturday,0


In [13]:
# Duplicate Check
duplicate_count = df.duplicated().sum()

duplicate_count

np.int64(0)

In [14]:
# Compare Column Consistency Across Files
reference_columns = set(pd.read_csv(raw_files[0]).columns)

column_check = []

for file in raw_files:
    temp_df = pd.read_csv(file)
    current_columns = set(temp_df.columns)
    
    column_check.append({
        "file_name": file.name,
        "columns_match_reference": current_columns == reference_columns,
        "missing_columns": list(reference_columns - current_columns),
        "extra_columns": list(current_columns - reference_columns)
    })

column_check_df = pd.DataFrame(column_check)

column_check_df

,file_name,columns_match_reference,missing_columns,extra_columns
0,annual_metropolitan_train_station_entries_fy_2...,True,[],[]
1,annual_metropolitan_train_station_entries_fy_2...,True,[],[]
2,annual_metropolitan_train_station_entries_fy_2...,True,[],[]
3,annual_metropolitan_train_station_entries_fy_2...,True,[],[]
4,annual_metropolitan_train_station_entries_fy_2...,True,[],[]
5,annual_metropolitan_train_station_entries_fy_2...,True,[],[]
6,annual_metropolitan_train_station_entries_fy_2...,True,[],[]


In [15]:
# Row Count Difference
inventory_df[["file_name", "rows", "columns"]].sort_values("rows", ascending=False)

,file_name,rows,columns
4,annual_metropolitan_train_station_entries_fy_2...,223,16
1,annual_metropolitan_train_station_entries_fy_2...,222,16
0,annual_metropolitan_train_station_entries_fy_2...,222,16
2,annual_metropolitan_train_station_entries_fy_2...,222,16
3,annual_metropolitan_train_station_entries_fy_2...,222,16
5,annual_metropolitan_train_station_entries_fy_2...,222,16
6,annual_metropolitan_train_station_entries_fy_2...,222,16


## Initial Observations

- All files contain 16 columns.
- Most yearly files contain 222 rows.
- The 2022–2023 dataset contains 223 rows and requires further investigation.
- The latest dataset loaded successfully.
- Further analysis is required to confirm whether column names and data types are consistent across all years.

## Next Steps

The next notebook will focus on:

- Cleaning column names
- Adding a financial year column
- Combining all annual files
- Investigating the additional row in 2022–2023
- Creating a clean master dataset for analysis